src/http.cc

```diff
@@ -826,11 +838,139 @@ HttpStateData::handle1xx(HttpReply *reply)
     // for similar reasons without a 1xx response.
 }
 
+/// A protocol-comparing predicate for STL search algorithms.
+/// Helps to match a protocol versioned name to a name without
+/// version (for example, 'websocket/1.2' should match to 'websocket' ).
+class MatchProtocol
+{
+public:
+    explicit MatchProtocol(const char *s, size_t len) :
+        reference(s),
+        referenceLen(len),
+        referenceBaseLen(static_cast<size_t>(-1))
+        {}
+
+    /// Operator to search in lists or vectors
+    bool operator() (const SBuf &checking) {
+        if (referenceIsVersioned()) {
+            const bool checkingIsVersioned = (checking.find('/') != SBuf::npos);
+            if (!checkingIsVersioned) {
+                assert(referenceBaseLen != static_cast<size_t>(-1));
+                return checking.caseCmp(reference, referenceBaseLen) == 0;
+            }
+        }
+        return checking.caseCmp(reference, referenceLen) == 0;
```

Comment:
> But it looks like MatchProtocol::operator() would return false in the second case

Yes this is intended.

The comment in JIRA  says:

> I believe that a (general) "name" matches any (specific) "name/X". And vice versa ...

This is means that we should completely drop version part when comparing.
But I still have my objections on "vice versa".

@rousskov 
Can you confirm which is the correct behaviour?

Reply:
AFAICT, the most natural behavior is to interpret missing version as "any version would do". In other words, `nameA` is `nameA/*`. That would result in the following matching rules:

1. nameA matches nameA
1. nameA matches nameA/X for any X
1. nameA/X matches nameA for any X

You seem to disagree that the above rules are correct, but I do not understand what your concern is so I cannot address it. That does not mean you are wrong!

Reply:
No problem for 1 and 2.

My  (not strong) objection is with the 3. But lets separate on two cases. Maybe we need to handle them with a different way.

### 1. While we are checking server upgrade response against client requested upgrade.

Probably should be "nameA/X matches nameA/X only".
If client requests the "protocol/x" means that he want to speak exactly this protocol version. A response of "protocol" may implies any release or different version of protocol than the X.

But of course websockets and probably other protocols which uses the upgrade feature  (eg TLS  or http2), are not relay on this header to negotiate the version.

### 2. While we are checking http_upgrade_request_protocols 

Imagine that the admin want to block protocol of version 13 and allow version 14 other:
`http_upgrade_request_protocols protocol/13 deny  all`
`http_upgrade_request_protocols protocol/14 allow  all`

Squid will block a request with "Upgrade: protocol/13" . What about if the request has just an "Upgrade: protocol" ? Which of the above rules should match?
I suppose none so in this case the rule 3 should not apply.

Reply:
> 2. While we are checking http_upgrade_request_protocols

Considering your example, I think that Squid should accept "Upgrade: protocol" against "http_upgrade_request_protocols protocol/14 allow all" rule and then generate and send to server "protocol/14" Upgrade header. I.e., when the client requests any protocol version, Squid should be able to select an applicable protocol version. For example, several versions may be allowed:

```
http_upgrade_request_protocols protocol/13 deny all
http_upgrade_request_protocols protocol/14 allow all
http_upgrade_request_protocols protocol/15 deny all
```
In this case, two rules match "Upgrade: protocol" Squid. Probably Squid should use the highest version, i.e., 15 in this case. What do you think?

Reply:
> 1. While we are checking server upgrade response against client requested upgrade.

However, as we [discussed](https://measurementfactory.atlassian.net/browse/SQUID-323?focusedCommentId=40212&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-40212) on JIRA, RFCs do not explicitly forbid matching server "protocol" response to the "protocol/x" request (at least I had not found such RFC rules). Do we need to introduce such a limitiation in Squid despite of this?

Reply:
> Probably Squid should use the highest version, i.e., 15 in this case. What do you think?

I would be very worried about tasking Squid with ordering protocol versions. A version token may not be a simple number! Is 15pre3 "greater" than 15rc2? [Rhetorical]

I think we should restrict ourselves to filtering (i.e. either allowing as-is or removing completely) each protocol that the client offers. Yes, it is tempting to change `foo` into `foo/15` in some cases, but that complex change should be left to adaptation services. It may be very difficult to design and use ACL-driven rules that correctly support _rewriting_ of the individual Upgrade header values. Let's start with something simpler.

I think that Christos' suggestion to consider two different evaluation contexts (http_upgrade_request_protocols evaluation and server response evaluation) is a great idea. Whatever the final outcome is, it is _easier_ to consider each of those contexts in isolation, at least as a starting point.

I will reorder them to use a more natural/evaluation order:

### 1. http_upgrade_request_protocols evaluation

> Imagine that the admin want to block protocol of version 13 and allow version 14 other:

```
http_upgrade_request_protocols protocol/13 deny all
http_upgrade_request_protocols protocol/14 allow all
```

> What about if the request has just an "Upgrade: protocol" ? Which of the above rules should match? I suppose none

s/match/be evaluated or considered or applied/

I agree that neither rule should be evaluated. When admin says "allow name/13", she is unlikely to mean "allow any version of the name protocol". We do not know whether she wants to allow bare "name" as well. It should be her responsibility to tell us (by adding a corresponding rule). It may be a good idea to warn her if she does not.

Again, it is tempting to say that we will rewrite "name" into "name/14" because a "name/14" rule matched, but then we would have to deal with accidents rewriting "name/14" into "name" and with questions like "how do I rewrite `name` into `name/13,name/14`?".

So, to summarize, in the http_upgrade_request_protocols context:

1. Rules with admin-provided versioned `name/X` protocol are applied to the same versioned `Upgrade:name/X` entries only.
1. Rules with admin-provided version-agnostic `name` protocol are applied to same version-agnostic `Upgrade:name` and any versioned `Upgrade:name/X` entries.

Side effect: An admin using any versioned `name/X` rules but wishing to allow version-agnostic `name` upgrades, must end name-related rules with a "name allow" rule. Without such a rule, no versioned `name/X` rules will be applied to a version-agnostic Upgrade:name protocol, and, with no applicable rules found, the upgrade to that protocol will be denied.

### 2. Server response evaluation

> If client requests the "protocol/x" means that he want to speak exactly this protocol version.

Agreed.

> A response of "protocol" may implies any release or different version of protocol than the X.

Difficult to say what a version-agnostic response implies in this context. The answer may depend on the protocol itself! We have three options:

1. Declare a match. Server knows better than Squid.
2. Declare a mismatch because we want to prevent some problems. What specific problems are we trying to prevent? Can we prevent them later (by modifying Squid) in a backward-compatible manner if we do not prevent them now? In other words, can we switch from this option 2 to option 3 below in a backward-compatible manner?
3. Ask admin.

If Squid does not know for sure, and the upgrade is deemed successful by the server, I doubt Squid should automatically declare a mismatch here. Thus, I would use option 1 or 3. However, if you have specific examples in favor of option 2, please discuss!

Option 3 is already kind of available because an admin can use http_reply_access to block specific Upgrade (HTTP/101) responses, with access to Squid-offered Upgrade protocol names and server-selected Upgrade protocols, right? If so, let's use option 1 because it is simpler.

There is a related problem here: Does a reordered `Upgrade:z,y` server response match the `Upgrade:x,y,z` Squid request? Based on the generic Upgrade description in the [RFC](https://tools.ietf.org/html/rfc7230#section-6.7), it can be a match. However, imagine enforcing that inside Squid when some of those entries are versioned and the server is adding or dropping versions -- restoring the right order to make sure that no extra protocol was sneaked in by the server might even become an NP problem!

I suggest that any successful server response is forwarded to the client (establishing a tunnel from Squid p.o.v.). If an admin does not like that default, they can either
* not enable Squid offers to tunnel via http_upgrade_request_protocols or
* block specific HTTP/101 responses from the server

In either case, this decision can be restricted to a specific server or using a specific protocol. We just need to check that HTTP/101 blocking works as expected (the client should get an error response and the to-server connection must be terminated when Squid denies an HTTP/101 response).

----

OK, so why not use the same simple "client knows best" behavior for Upgrade requests and avoid adding http_upgrade_request_protocols?! After all, we do have http_access and request_header_access rules that can be used to police Upgrade requests! AFAICT, we have to add http_upgrade_request_protocols because we do not want to start allowing upgrades _by default_, and we cannot force admins that do not want such new default to rewrite their existing rules to start blocking upgrades.

Reply:
Just to clarify what we want to implement.

We concluded what we have two cases:
1) http_upgrade_request_protocols evaluation
2) Server response evaluation

which are handled with a different way. In the first case we want to allow admin to have a control over requested upgrate protocol version. In the second case the server knows better which release to select so we do not want to involved here.

Now lets assume the case the client requests "protocol/12" and the server responds with "protocol/15" or "protocol/10". Should the Server response evaluation allow it?
And what in the case the http_upgrade_request_protocols evaluation has deny the "protocol/15" and "protocol/10" while the request sent?
To make a more useful example lets assume that the protocol ls the "TLS" and the admin wants to allow only upgrading to secure TLS/1.2 and not previous TLS  releases.
It is not clear what the release means. It maybe means an "up-to" or "at-last-to" or anything else.

My understanding is that:
 - We should to give to the admin a full control over the requested version 
 - But do not take at all  in account the requested and selected (from server) upgrading protocol releases on server response evaluation. Only check the base protocols. If the admin wants it can deny the response for example using the http_reply_access acls.

Am I correct?

Reply:
Fixed at 263f3d0d
Resolved at d6834a56
Fixed again at c716c76b